File: gen_rabani2.m
% Andrew Stannard 27/06/19, Rabani model in Matlab
clear variables; close all
cm = colormap([0 0 0; 
               1 1 1; 
               1 0.5 0]);
mus = 0:0.2:6;
kTs = 0:0.2:6;

parfor mu_cnt = 1:length(mus)
    for kT_cnt = 1:length(kTs)
        mu = mus(mu_cnt);
        kT = kTs(kT_cnt);
        disp([mu, kT])
        L = 128; N = L^2; % system length and size
        MCS = 450; % total number of Monte Carlo steps
        MR = 30; % mobiity ratio
        C = 0.30; % fractional coverage of nanoparticless
        B = 1/kT; % thermal energy and invrse thermal energy
        e_nl = 1.5; % nanoparticle-liquid interaction energy
        e_nn = 2.0; % nanoparticle-nanoparticle interaction energy
        I = randperm(N,round(C*N)); % random initial nanoparticle positions
        NP = zeros(L); NP(I) = 1; % adding nanoparticles to nanoparticle array
        LQ = ones(L); LQ(I) = 0;  % removing corresponding liquid from liquid array

        % Random number arrays for the diffusion loops
        X1 = ceil(L*rand(MCS, N)); Y1 = ceil(L*rand(MCS, N));
        X2 = ceil(L*rand(MCS, N*MR)); Y2 = ceil(L*rand(MCS, N*MR));
      
        % Nearest and next nearest neighbour arrays with periodic boundaries
        XP11 = X1+1-L*(X1==L); 
        XP21 = X1+2-L*(X1>=L-1);
        YP11 = Y1+1-L*(Y1==L); 
        YP21 = Y1+2-L*(Y1>=L-1);
        XM11 = X1-1+L*(X1==1);         
        XM21 = X1-2+L*(X1<=2);
        YM11 = Y1-1+L*(Y1==1); 
        YM21 = Y1-2+L*(Y1<=2);
        
        XP12 = X2+1-L*(X2==L); 
        XP22 = X2+2-L*(X2>=L-1);
        YP12 = Y2+1-L*(Y2==L); 
        YP22 = Y2+2-L*(Y2>=L-1);
        XM12 = X2-1+L*(X2==1); 
        XM22 = X2-2+L*(X2<=2);
        YM12 = Y2-1+L*(Y2==1); 
        YM22 = Y2-2+L*(Y2<=2);

        R1 = rand(MCS, N,1); % random number array for Metropolis acceptance
        R2 = rand(MCS, N*MR,1); % random number array for Metropolis acceptance

        for m = 1 : MCS % main Monte Carlo loop
            for i = 1 : N % start of evaporation/condensation loop       
                if NP(X1(m, i),Y1(m, i)) == 0     
                    if LQ(X1(m, i),Y1(m, i)) == 0

                        % change in energy if condensation occurs
                        dE = -(LQ(XP11(m, i),Y1(m, i))+LQ(XM11(m, i),Y1(m, i))+LQ(X1(m, i),YP11(m, i))+LQ(X1(m, i),YM11(m, i)))...
                            -e_nl*(NP(XP11(m, i),Y1(m, i))+NP(XM11(m, i),Y1(m, i))+NP(X1(m, i),YP11(m, i))+NP(X1(m, i),YM11(m, i))) + mu;

                        if R1(m, i) < exp(-B*dE) % Metropolis acceptance
                            LQ(X1(m, i),Y1(m, i)) = 1; % condensation
                        end

                    else % i.e. if LQ(x,Y2(m, i)) == 1

                        % change in energy if evaporation occurs
                        dE = (LQ(XP11(m, i),Y1(m, i))+LQ(XM11(m, i),Y1(m, i))+LQ(X1(m, i),YP11(m, i))+LQ(X1(m, i),YM11(m, i)))...
                            +e_nl*(NP(XP11(m, i),Y1(m, i))+NP(XM11(m, i),Y1(m, i))+NP(X1(m, i),YP11(m, i))+NP(X1(m, i),YM11(m, i))) - mu;    
                        if R1(m, i) < exp(-B*dE) % Metropolis acceptance
                            LQ(X1(m, i),Y1(m, i)) = 0; % evaporation
                        end              
                    end
                end        
            end % end of evaporation/condensation loop

            % random number array for nanoparticle movement direction
            D = ceil(4*rand(N*MR,1)); % 1 = left, 2 = right, 3 = down, 4 = up

            for i = 1 : N * MR % start of nanoparticle diffusion loop   
                if NP(X2(m, i),Y2(m, i)) == 1            
                    if (D(i) == 1) && (LQ(XM12(m, i),Y2(m, i)) == 1)

                        % change in energy if nanoparticle moves left
                        dE = -(LQ(X2(m, i),YM12(m, i)) + LQ(XP12(m, i),Y2(m, i)) + LQ(X2(m, i),YP12(m, i)) ...
                            - LQ(XM22(m, i),Y2(m, i)) - LQ(XM12(m, i),YP12(m, i)) - LQ(XM12(m, i),YM12(m, i))) ...
                            -e_nl * (LQ(XM22(m, i),Y2(m, i)) + LQ(XM12(m, i),YP12(m, i)) + LQ(XM12(m, i),YM12(m, i)) ...
                            - LQ(X2(m, i),YM12(m, i)) - LQ(XP12(m, i),Y2(m, i)) - LQ(X2(m, i),YP12(m, i))) ...
                            -e_nl * (NP(X2(m, i),YM12(m, i)) + NP(XP12(m, i),Y2(m, i)) + NP(X2(m, i),YP12(m, i)) ...
                            - NP(XM22(m, i),Y2(m, i)) - NP(XM12(m, i),YP12(m, i)) - NP(XM12(m, i),YM12(m, i))) ...
                            -e_nn * (NP(XM22(m, i),Y2(m, i)) + NP(XM12(m, i),YP12(m, i)) + NP(XM12(m, i),YM12(m, i)) ...
                            - NP(X2(m, i),YM12(m, i)) - NP(XP12(m, i),Y2(m, i)) - NP(X2(m, i),YP12(m, i)));

                        if R2(m, i) < exp(-B*dE) % Metropolis acceptance
                            NP(XM12(m, i),Y2(m, i)) = 1; NP(X2(m, i),Y2(m, i)) = 0; % nanoparticle moves left
                            LQ(X2(m, i),Y2(m, i)) = 1; LQ(XM12(m, i),Y2(m, i)) = 0; % liquid moves right
                        end

                    elseif (D(i) == 2) && (LQ(XP12(m, i),Y2(m, i)) == 1)

                        % change in energy if nanoparticle moves right
                        dE = -(LQ(X2(m, i),YM12(m, i)) + LQ(XP12(m, i),Y2(m, i)) + LQ(X2(m, i),YP12(m, i)) ...
                            - LQ(XP22(m, i),Y2(m, i)) - LQ(XP12(m, i),YP12(m, i)) - LQ(XP12(m, i),YM12(m, i))) ...
                            -e_nl * (LQ(XP22(m, i),Y2(m, i)) + LQ(XP12(m, i),YP12(m, i)) + LQ(XP12(m, i),YM12(m, i)) ...
                            - LQ(X2(m, i),YM12(m, i)) - LQ(XM12(m, i),Y2(m, i)) - LQ(X2(m, i),YP12(m, i))) ...
                            -e_nl * (NP(X2(m, i),YM12(m, i)) + NP(XM12(m, i),Y2(m, i)) + NP(X2(m, i),YP12(m, i)) ...
                            - NP(XP22(m, i),Y2(m, i))-NP(XP12(m, i),YP12(m, i))-NP(XP12(m, i),YM12(m, i))) ...
                            -e_nn * (NP(XP22(m, i),Y2(m, i)) + NP(XP12(m, i),YP12(m, i)) + NP(XP12(m, i),YM12(m, i)) ...
                            - NP(X2(m, i),YM12(m, i)) - NP(XM12(m, i),Y2(m, i)) - NP(X2(m, i),YP12(m, i)));

                        if R2(m, i) < exp(-B*dE) % Metropolis acceptance
                            NP(XP12(m, i),Y2(m, i)) = 1; NP(X2(m, i),Y2(m, i)) = 0; % nanoparticle moves right
                            LQ(X2(m, i),Y2(m, i)) = 1; LQ(XP12(m, i),Y2(m, i)) = 0; % liquid moves left
                        end

                    elseif (D(i) == 3) && (LQ(X2(m, i),YM12(m, i)) == 1)

                        % change in energy if nanoparticle moves down
                        dE = -(LQ(XM12(m, i),Y2(m, i)) + LQ(X2(m, i),YP12(m, i)) + LQ(XP12(m, i),Y2(m, i)) ...
                            - LQ(XM12(m, i),YM12(m, i)) - LQ(X2(m, i),YM22(m, i)) - LQ(XP12(m, i),YM12(m, i))) ...
                            -e_nl * (LQ(XM12(m, i),YM12(m, i)) + LQ(X2(m, i),YM22(m, i)) + LQ(XP12(m, i),YM12(m, i)) ...
                            - LQ(XM12(m, i),Y2(m, i)) - LQ(X2(m, i),YP12(m, i)) - LQ(XP12(m, i),Y2(m, i))) ...
                            -e_nl * (NP(XM12(m, i),Y2(m, i)) + NP(X2(m, i),YP12(m, i)) + NP(XP12(m, i),Y2(m, i)) ...
                            - NP(XM12(m, i),YM12(m, i)) - NP(X2(m, i),YM22(m, i)) - NP(XP12(m, i),YM12(m, i))) ...
                            -e_nn * (NP(XM12(m, i),YM12(m, i)) + NP(X2(m, i),YM22(m, i)) + NP(XP12(m, i),YM12(m, i)) ...
                            - NP(XM12(m, i),Y2(m, i)) - NP(X2(m, i),YP12(m, i)) - NP(XP12(m, i),Y2(m, i)));

                        if R2(m, i) < exp(-B*dE) % Metropolis acceptance
                            NP(X2(m, i),YM12(m, i)) = 1; NP(X2(m, i),Y2(m, i)) = 0; % nanoparticle moves down
                            LQ(X2(m, i),Y2(m, i)) = 1; LQ(X2(m, i),YM12(m, i)) = 0; % liquid moves up 
                        end

                    elseif (D(i) == 4) && (LQ(X2(m, i),YP12(m, i)) == 1)

                        % change in energy if nanoparticle moves up
                        dE = -(LQ(XM12(m, i),Y2(m, i)) + LQ(X2(m, i),YM12(m, i)) + LQ(XP12(m, i),Y2(m, i)) ...
                            - LQ(XM12(m, i),YP12(m, i)) - LQ(X2(m, i),YP22(m, i)) - LQ(XP12(m, i),YP12(m, i))) ...
                            -e_nl * (LQ(XM12(m, i),YP12(m, i)) + LQ(X2(m, i),YP22(m, i)) + LQ(XP12(m, i),YP12(m, i)) ...
                            - LQ(XM12(m, i),Y2(m, i)) - LQ(X2(m, i),YM12(m, i)) - LQ(XP12(m, i),Y2(m, i))) ...
                            -e_nl * (NP(XM12(m, i),Y2(m, i)) + NP(X2(m, i),YM12(m, i)) + NP(XP12(m, i),Y2(m, i)) ...
                            - NP(XM12(m, i),YP12(m, i)) - NP(X2(m, i),YP22(m, i)) - NP(XP12(m, i),YP12(m, i))) ...
                            -e_nn * (NP(XM12(m, i),YP12(m, i)) + NP(X2(m, i),YP22(m, i)) + NP(XP12(m, i),YP12(m, i)) ...
                            - NP(XM12(m, i),Y2(m, i)) - NP(X2(m, i),YM12(m, i)) - NP(XP12(m, i),Y2(m, i)));

                        if R2(m, i) < exp(-B*dE) % Metropolis acceptance
                            NP(X2(m, i),YP12(m, i)) = 1; NP(X2(m, i),Y2(m, i)) = 0; % nanoparticle moves up
                            LQ(X2(m, i),Y2(m, i)) = 1; LQ(X2(m, i),YP12(m, i)) = 0; % liquid moves down 
                        end
                    end
                end
            end
            
        end
        fname = "ImagesMatlabNew2/rabani_kT=" + num2str(kT, '%0.2f') + "mu="  + num2str(mu, '%0.2f') + ".png";
        
        img = (2*NP+LQ)+1;
        imwrite(img, cm, fname);
        
    end
end

File: gen_rabani2.py
import numpy as np
from matplotlib import colors, pyplot as plt
from numba import jit, njit, prange

@jit(nopython=True)
def rabani():
    L = 128  # System length
    N = L ** 2  # System size
    MCS = 30  # Total num. of Monte Carlo steps
    MR = 30  # Mobility ratio
    C = 0.30  # Fractional coverage of nanoparticles
    kT = 0.25  # Thermal energy
    B = 1 / kT  # Inverse thermal energy
    e_nl = 1.5  # Interaction energy nanoparticle-liquid
    e_nn = 2.0  # Interaction energy nanoparticle-nanoparticle
    mu = 2.5  # Liquid chemical potential

    # Random initial nanoparticle positions
    I = np.random.choice(N, int(C * N), replace=False)
    NP = np.zeros((N,))
    NP[I] = 1
    NP = NP.reshape((L, L))
    LQ = np.abs(1 - NP)

    for m in range(MCS):
        # Random position arrays for the evaporation/condensation loop
        X = np.ceil(L * np.random.random((N,))) - 1
        Y = np.ceil(L * np.random.random((N,))) - 1

        # Nearest and next nearest neighbour arrays with periodic boundaries
        XP1 = X + 1 - L * (X == (L - 1))
        XP2 = X + 2 - L * (X >= (L - 2))
        YP1 = Y + 1 - L * (Y == (L - 1))
        YP2 = Y + 2 - L * (Y >= (L - 2))
        XM1 = X - 1 + L * (X == 0)
        XM2 = X - 2 + L * (X <= 1)
        YM1 = Y - 1 + L * (Y == 0)
        YM2 = Y - 2 + L * (Y <= 1)

        # Random number array for Metropolis acceptance
        R = np.random.random((N,))

        for i in range(N):  # Start of evaporation/condensation loop

            x = int(X[i])
            xp1 = int(XP1[i])
            xp2 = int(XP2[i])
            xm1 = int(XM1[i])
            xm2 = int(XM2[i])
            y = int(Y[i])
            yp1 = int(YP1[i])
            yp2 = int(YP2[i])
            ym1 = int(YM1[i])
            ym2 = int(YM2[i])
            r = R[i]
            # print(r, R[i])

            if NP[x, y] == 0:
                if LQ[x, y] == 0:
                    # change in energy if condensation occurs
                    dE = -(LQ[xp1, y] + LQ[xm1, y] + LQ[x, yp1] + LQ[x, ym1]) - e_nl * (
                            NP[xp1, y] + NP[xm1, y] + NP[x, yp1] + NP[x, ym1]) + mu

                    if r < np.exp(-B * dE):  # Metropolis acceptance
                        LQ[x, y] = 1  # condensation
                else:
                    # change in energy if evaporation occurs
                    dE = (LQ[xp1, y] + LQ[xm1, y] + LQ[x, yp1] + LQ[x, ym1]) + e_nl * (
                            NP[xp1, y] + NP[xm1, y] + NP[x, yp1] + NP[x, ym1]) - mu
                    if r < np.exp(-B * dE):  # Metropolis acceptance
                        LQ[x, y] = 0  # evaporation

        # Random number arrays for the nanoparticle diffusion loop
        X = np.ceil(L * np.random.random((N * MR,))) - 1
        Y = np.ceil(L * np.random.random((N * MR,))) - 1

        # Nearest and next nearest neighbour arrays with periodic boundaries
        XP1 = X + 1 - L * (X == (L - 1))
        XP2 = X + 2 - L * (X >= (L - 2))
        YP1 = Y + 1 - L * (Y == (L - 1))
        YP2 = Y + 2 - L * (Y >= (L - 2))
        XM1 = X - 1 + L * (X == 0)
        XM2 = X - 2 + L * (X <= 1)
        YM1 = Y - 1 + L * (Y == 0)
        YM2 = Y - 2 + L * (Y <= 1)

        R = np.random.random((N * MR,))  # random number array for Metropolis acceptance

        # Random number array for nanoparticle movement direction
        D = np.ceil(4 * np.random.random((N * MR,)))  # 1 = left, 2 = right, 3 = down, 4 = up

        for i in range(N * MR):  # start of nanoparticle diffusion loop
            x = int(X[i])
            xp1 = int(XP1[i])
            xp2 = int(XP2[i])
            xm1 = int(XM1[i])
            xm2 = int(XM2[i])
            y = int(Y[i])
            yp1 = int(YP1[i])
            yp2 = int(YP2[i])
            ym1 = int(YM1[i])
            ym2 = int(YM2[i])
            r = R[i]
            d = int(D[i])

            if NP[x, y] == 1:

                if (d == 1) and (LQ[xm1, y] == 1):

                    # change in energy if nanoparticle moves left
                    dE = -(LQ[x, ym1] + LQ[xp1, y] + LQ[x, yp1] - LQ[xm2, y] - LQ[xm1, yp1] - LQ[xm1, ym1]) - e_nl * (
                            LQ[xm2, y] + LQ[xm1, yp1] + LQ[xm1, ym1] - LQ[x, ym1] - LQ[xp1, y] - LQ[x, yp1]) - e_nl * (
                                 NP[x, ym1] + NP[xp1, y] + NP[x, yp1] - NP[xm2, y] - NP[xm1, yp1] - NP[
                             xm1, ym1]) - e_nn * (
                                 NP[xm2, y] + NP[xm1, yp1] + NP[xm1, ym1] - NP[x, ym1] - NP[xp1, y] - NP[x, yp1])

                    if r < np.exp(-B * dE):  # Metropolis acceptance
                        NP[xm1, y] = 1
                        NP[x, y] = 0  # nanoparticle moves left
                        LQ[x, y] = 1
                        LQ[xm1, y] = 0  # liquid moves right

                elif (d == 2) and (LQ[xp1, y] == 1):
                    # Change in energy if nanoparticle moves right
                    dE = -(LQ[x, ym1] + LQ[xp1, y] + LQ[x, yp1] - LQ[xp2, y] - LQ[xp1, yp1] - LQ[xp1, ym1]) - e_nl * (
                            LQ[xp2, y] + LQ[xp1, yp1] + LQ[xp1, ym1] - LQ[x, ym1] - LQ[xm1, y] - LQ[x, yp1]) - e_nl * (
                                 NP[x, ym1] + NP[xm1, y] + NP[x, yp1] - NP[xp2, y] - NP[xp1, yp1] - NP[
                             xp1, ym1]) - e_nn * (
                                    NP[xp2, y] + NP[xp1, yp1] + NP[xp1, ym1] - NP[x, ym1] - NP[xm1, y] - NP[x, yp1])

                    if r < np.exp(-B * dE):  # Metropolis acceptance
                        NP[xp1, y] = 1
                        NP[x, y] = 0  # Nanoparticle moves right
                        LQ[x, y] = 1
                        LQ[xp1, y] = 0  # Liquid moves left

                elif (d == 3) and (LQ[x, ym1] == 1):
                    # change in energy if nanoparticle moves down
                    dE = -(LQ[xm1, y] + LQ[x, yp1] + LQ[xp1, y] - LQ[xm1, ym1] - LQ[x, ym2] - LQ[xp1, ym1]) - e_nl * (
                            LQ[xm1, ym1] + LQ[x, ym2] + LQ[xp1, ym1] - LQ[xm1, y] - LQ[x, yp1] - LQ[xp1, y]) - e_nl * (
                                 NP[xm1, y] + NP[x, yp1] + NP[xp1, y] - NP[xm1, ym1] - NP[x, ym2] - NP[
                             xp1, ym1]) - e_nn * (
                                 NP[xm1, ym1] + NP[x, ym2] + NP[xp1, ym1] - NP[xm1, y] - NP[x, yp1] - NP[xp1, y])

                    if r < np.exp(-B * dE):  # Metropolis acceptance
                        NP[x, ym1] = 1
                        NP[x, y] = 0  # Nanoparticle moves down
                        LQ[x, y] = 1
                        LQ[x, ym1] = 0  # Liquid moves up

                elif (d == 4) and (LQ[x, yp1] == 1):
                    # Change in energy if nanoparticle moves up
                    dE = -(LQ[xm1, y] + LQ[x, ym1] + LQ[xp1, y] - LQ[xm1, yp1] - LQ[x, yp2] - LQ[xp1, yp1]) - e_nl * (
                            LQ[xm1, yp1] + LQ[x, yp2] + LQ[xp1, yp1] - LQ[xm1, y] - LQ[x, ym1] - LQ[xp1, y]) - e_nl * (
                                 NP[xm1, y] + NP[x, ym1] + NP[xp1, y] - NP[xm1, yp1] - NP[x, yp2] - NP[
                             xp1, yp1]) - e_nn * (
                                 NP[xm1, yp1] + NP[x, yp2] + NP[xp1, yp1] - NP[xm1, y] - NP[x, ym1] - NP[xp1, y])

                    if r < np.exp(-B * dE):  # Metropolis acceptance
                        NP[x, yp1] = 1
                        NP[x, y] = 0  # Nanoparticle moves up
                        LQ[x, y] = 1
                        LQ[x, yp1] = 0  # Liquid moves down

        # Check if equilibrium reached
        lastrun = 2 * NP + LQ

# if __name__ == '__main__':
#     cmap = colors.ListedColormap(["black", "white", "orange"])
#     boundaries = [0, 0.5, 1]
#     norm = colors.BoundaryNorm(boundaries, cmap.N, clip=True)
#     plt.imshow(lastrun, cmap=cmap)
#     plt.axis("square")
#     plt.axis("off")
#     plt.pause(0.001)

